core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -251,8 +251,9 @@ class Partition(val topicPartition: TopicPartition,
 
   // lock to prevent the follower replica log update while checking if the log dir could be replaced with future log.
   private val futureLogLock = new Object()
-  private var zkVersion: Int = LeaderAndIsr.initialZKVersion
-  @volatile private var leaderEpoch: Int = LeaderAndIsr.initialLeaderEpoch - 1
+  // The current epoch for the partition for KRaft controllers. The current ZK version for the legacy controllers.
+  private var partitionEpoch: Int = LeaderAndIsr.InitialPartitionEpoch
```

Comment:
Do you know why this field is not volatile like `leaderEpoch`.

Reply:
It seems that `partitionEpoch` is always accessed within the `leaderIsrUpdateLock` lock whereas `leaderEpoch` is also accessed by the `ReplicaManager`. Do you think that we should make it volatile to be consistent?

Reply:
Either that or we should try to document the expectation that we have the lock. One thing that makes me a little uncomfortable in regard to these volatile fields is any expectation of consistency between them. Maybe a potential refactor for a subsequent patch could be to pull these fields into `PartitionState` so that the state can be updated atomically. What do you think?

Reply:
Yeah, that makes sense. I will make it volatile for now. I have filled https://issues.apache.org/jira/browse/KAFKA-13844 for the refactor. I will take care of this as a follow up.